Marlin/src/HAL/HAL_LPC1768/pinmapping.h

```diff
@@ -24,13 +24,147 @@
+const HAL_PIN_TYPE P1_9  = LPC1768_PIN(1,  9);
+const HAL_PIN_TYPE P1_10 = LPC1768_PIN(1, 10);
+const HAL_PIN_TYPE P1_11 = LPC1768_PIN(1, 11);
+const HAL_PIN_TYPE P1_12 = LPC1768_PIN(1, 12);
+const HAL_PIN_TYPE P1_13 = LPC1768_PIN(1, 13);
+const HAL_PIN_TYPE P1_14 = LPC1768_PIN(1, 14);
+const HAL_PIN_TYPE P1_15 = LPC1768_PIN(1, 15);
+const HAL_PIN_TYPE P1_16 = LPC1768_PIN(1, 16);
+const HAL_PIN_TYPE P1_17 = LPC1768_PIN(1, 17);
+const HAL_PIN_TYPE P1_18 = LPC1768_PIN(1, 18, PIN_FEATURE_PWM);
+const HAL_PIN_TYPE P1_19 = LPC1768_PIN(1, 19);
+const HAL_PIN_TYPE P1_20 = LPC1768_PIN(1, 20, PIN_FEATURE_PWM);
+const HAL_PIN_TYPE P1_21 = LPC1768_PIN(1, 21, PIN_FEATURE_PWM);
+const HAL_PIN_TYPE P1_22 = LPC1768_PIN(1, 22);
+const HAL_PIN_TYPE P1_23 = LPC1768_PIN(1, 23, PIN_FEATURE_PWM);
+const HAL_PIN_TYPE P1_24 = LPC1768_PIN(1, 24, PIN_FEATURE_PWM);
+const HAL_PIN_TYPE P1_25 = LPC1768_PIN(1, 25);
+const HAL_PIN_TYPE P1_26 = LPC1768_PIN(1, 26, PIN_FEATURE_PWM);
+const HAL_PIN_TYPE P1_27 = LPC1768_PIN(1, 27);
+const HAL_PIN_TYPE P1_28 = LPC1768_PIN(1, 28);
+const HAL_PIN_TYPE P1_29 = LPC1768_PIN(1, 29);
+const HAL_PIN_TYPE P1_30 = LPC1768_PIN(1, 30, PIN_FEATURE_ADC(4));
+const HAL_PIN_TYPE P1_31 = LPC1768_PIN(1, 31, PIN_FEATURE_ADC(5));
+const HAL_PIN_TYPE P2_0  = LPC1768_PIN(2,  0, PIN_FEATURE_INTERRUPT | PIN_FEATURE_PWM);
+const HAL_PIN_TYPE P2_1  = LPC1768_PIN(2,  1, PIN_FEATURE_INTERRUPT | PIN_FEATURE_PWM);
+const HAL_PIN_TYPE P2_2  = LPC1768_PIN(2,  2, PIN_FEATURE_INTERRUPT | PIN_FEATURE_PWM);
+const HAL_PIN_TYPE P2_3  = LPC1768_PIN(2,  3, PIN_FEATURE_INTERRUPT | PIN_FEATURE_PWM);
+const HAL_PIN_TYPE P2_4  = LPC1768_PIN(2,  4, PIN_FEATURE_INTERRUPT | PIN_FEATURE_PWM);
+const HAL_PIN_TYPE P2_5  = LPC1768_PIN(2,  5, PIN_FEATURE_INTERRUPT | PIN_FEATURE_PWM);
+const HAL_PIN_TYPE P2_6  = LPC1768_PIN(2,  6, PIN_FEATURE_INTERRUPT);
+const HAL_PIN_TYPE P2_7  = LPC1768_PIN(2,  7, PIN_FEATURE_INTERRUPT);
+const HAL_PIN_TYPE P2_8  = LPC1768_PIN(2,  8, PIN_FEATURE_INTERRUPT);
+const HAL_PIN_TYPE P2_9  = LPC1768_PIN(2,  9);
+const HAL_PIN_TYPE P2_10 = LPC1768_PIN(2, 10);
+const HAL_PIN_TYPE P2_11 = LPC1768_PIN(2, 11, PIN_FEATURE_INTERRUPT);
+const HAL_PIN_TYPE P2_12 = LPC1768_PIN(2, 12);
+const HAL_PIN_TYPE P2_13 = LPC1768_PIN(2, 13, PIN_FEATURE_INTERRUPT);
+const HAL_PIN_TYPE P3_25 = LPC1768_PIN(3, 25);
+const HAL_PIN_TYPE P3_26 = LPC1768_PIN(3, 26);
+const HAL_PIN_TYPE P4_28 = LPC1768_PIN(4, 28);
+const HAL_PIN_TYPE P4_29 = LPC1768_PIN(4, 29);
+
+constexpr bool VALID_PIN(HAL_PIN_TYPE p) {
+  return ((LPC1768_PIN_PORT(p) == 0 && LPC1768_PIN_PIN(p) <= 30) ? true :
+          (LPC1768_PIN_PORT(p) == 1 && LPC1768_PIN_PIN(p) <= 31) ? true :
+          (LPC1768_PIN_PORT(p) == 2 && LPC1768_PIN_PIN(p) <= 13) ? true :
+          (LPC1768_PIN_PORT(p) == 3 && LPC1768_PIN_PIN(p) >= 25 && LPC1768_PIN_PIN(p) <= 26) ? true :
+          (LPC1768_PIN_PORT(p) == 4 && LPC1768_PIN_PIN(p) >= 28 && LPC1768_PIN_PIN(p) <= 29) ? true :
+          false);
+}
```

Comment:
You can always replace `x = condition ? : true : false' with `x = condition`. So this could be:
```cpp
  return (   (LPC1768_PIN_PORT(p) == 0 && LPC1768_PIN_PIN(p) <= 30)
          || (LPC1768_PIN_PORT(p) == 1 && LPC1768_PIN_PIN(p) <= 31)
          || (LPC1768_PIN_PORT(p) == 2 && LPC1768_PIN_PIN(p) <= 13)
          || (LPC1768_PIN_PORT(p) == 3 && WITHIN(LPC1768_PIN_PIN(p), 25, 26)
          || (LPC1768_PIN_PORT(p) == 4 && WITHIN(LPC1768_PIN_PIN(p), 28, 29)
         );
```
…but since Repetition Is Evil, this is the best:
```cpp
  switch (LPC1768_PIN_PORT(p)) {
    case 0: return LPC1768_PIN_PIN(p) <= 30;
    case 1: return LPC1768_PIN_PIN(p) <= 31;
    case 2: return LPC1768_PIN_PIN(p) <= 13;
    case 3: return WITHIN(LPC1768_PIN_PIN(p), 25, 26);
    case 4: return WITHIN(LPC1768_PIN_PIN(p), 28, 29);
  }
```

Reply:
It turns out that the switch statement version won't work. `constexpr` functions on C++11 can only be a single line return statement. Since `VALID_PIN` is used by other `constexpr` functions, then it also has to be `constexpr`.